src/client/opamClient.ml

```diff
@@ -202,9 +202,25 @@ let upgrade_t
         else OpamPackage.packages_of_names t.installed requested
       in
       let latest =
-        OpamPackage.Name.Set.fold (fun name acc ->
-            OpamPackage.Set.add (OpamPackage.max_version t.packages name) acc)
-          (OpamPackage.names_of_packages to_check)
+        OpamPackage.Set.fold (fun pkg acc ->
+            let name = OpamPackage.name pkg in
+            let pkgs = OpamPackage.packages_of_name t.packages name in
+            let latest =
+              OpamPackage.Set.fold (fun pkg latest ->
+                  if OpamPackage.compare latest pkg < 0 then
+                    let opam = OpamPackage.Map.find pkg t.opams in
+                    let avoid_version =
+                      List.exists (function
+                          | Pkgflag_AvoidVersion | Pkgflag_Deprecated -> true
+                          | _ -> false)
```

Comment:
We should expand the default case, like that if we add another flag and this part is concerned there is an highlight.